sw/device/tests/sim_dv/gpio_test.c

```diff
@@ -0,0 +1,295 @@
+ * functionality is moved into pinmux.
+ *
+ * In the input direction, the external testbench sends the following pattern:
+ * 1. Walk a 1 in 'temperature' pattern (0001, 0011, 0111, 1111, 1110, 100, ..)
+ *
+ * Both, rising and falling edges are configured for generating an interrupt. As
+ * each pin rises or falls, the SW checks the interrupt, status and `data_in`
+ * for correctness.
+ */
+
+/**
+ * Initializes PLIC and enables all GPIO interrupts.
+ */
+static void plic_init_with_irqs(mmio_region_t base_addr, dif_plic_t *plic) {
+  LOG_INFO("Initializing the PLIC.");
+
+  CHECK(dif_plic_init(base_addr, plic) == kDifPlicOk, "dif_plic_init failed");
+
+  for (uint32_t i = 0; i < kNumGpios; ++i) {
+    dif_plic_irq_id_t plic_irq_id = i + kTopEarlgreyPlicIrqIdGpioGpio0;
+
+    // Enable GPIO interrupts at PLIC as edge triggered.
+    CHECK(dif_plic_irq_trigger_type_set(plic, plic_irq_id, kDifPlicEnable) ==
+              kDifPlicOk,
+          "dif_plic_irq_trigger_type_set failed");
+
+    // Set the priority of GPIO interrupts at PLIC to be >=1
+    CHECK(dif_plic_irq_priority_set(plic, plic_irq_id, 0x1) == kDifPlicOk,
+          "dif_plic_irq_priority_set failed");
+
+    // Enable all GPIO interrupts at the PLIC.
+    CHECK(
+        dif_plic_irq_enable_set(plic, plic_irq_id, kTopEarlgreyPlicTargetIbex0,
+                                kDifPlicEnable) == kDifPlicOk,
+        "dif_plic_irq_enable_set failed");
+  }
+
+  // Set the threshold for the Ibex to 0.
+  CHECK(dif_plic_target_threshold_set(plic, kTopEarlgreyPlicTargetIbex0, 0x0) ==
+            kDifPlicOk,
+        "dif_plic_target_threshold_set failed");
+}
+
+/**
+ * Runs the GPIO output test.
+ *
+ * Walks a 1 over the GPIO pins, followed by walking a 0.
+ * The external testbench checks the GPIO values for correctness.
+ */
+static void gpio_output_test(const dif_gpio_t *gpio) {
+  LOG_INFO("Starting GPIO output test");
+
+  // Set the GPIOs to be in output mode.
+  CHECK(dif_gpio_output_mode_all_set(gpio, kGpiosAllowedMask) == kDifGpioOk,
+        "dif_gpio_output_mode_all_set failed");
+
+  // Walk 1s
+  for (uint32_t i = 0; i < kNumGpios; ++i) {
+    uint32_t gpio_val = 1 << i;
+    CHECK(dif_gpio_all_write(gpio, gpio_val) == kDifGpioOk,
+          "dif_gpio_all_write failed");
+
+    // Read GPIO_IN to confirm what we wrote.
+    uint32_t read_val;
+    CHECK(dif_gpio_all_read(gpio, &read_val) == kDifGpioOk,
+          "dif_gpio_all_read failed");
+
+    // Check written and read val for correctness.
+    // Though we try to set all available GPIOs, only the ones that are exposed
+    // as chip IOs can be read back. So we mask the values with
+    // `kChipGpiosMask`.
+    gpio_val &= kChipGpiosMask;
+    read_val &= kChipGpiosMask;
+    CHECK(gpio_val == read_val, "GPIOs mismatched (written = %x, read = %x)",
+          gpio_val, read_val);
+  }
+
+  // Write all 0s to the GPIOs.
+  CHECK(dif_gpio_all_write(gpio, ~kGpiosMask) == kDifGpioOk,
```

Comment:
Is there anything that actually checks that the write goes through? I assume that happens on the SV side?

Reply:
Yes, the SV test sequence checks the correctness of the pattern (see updates to the file `chip_sw_gpio_test_vseq` in this PR if interested).